File: SonarMinefield/app/src/main/java/br/odb/sonarminefield/GameSession.java
/**
 *
 */
package br.odb.sonarminefield;

import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Random;

/**
 * @author monty
 *
 */
public class GameSession {

    public static final int WIDTH = 10;
    public static final int HEIGHT = 15;

    public static final int POSITION_MINE = 9;
    public static final int POSITION_COVERED = 10;
    public static final int POSITION_MINE_POKED = 11;
    public static final int POSITION_FLAGGED = 12;
    public static final int POSITION_BLANK = 0;

    public static final int GAMESTATE_NOTSTARTED = 0;
    public static final int GAMESTATE_STARTED = 1;
    public static final int GAMESTATE_FININSHED = 2;
    public static final int GAMESTATE_GAMEOVER = 3;

    private int[][] map;
    private boolean[][] covered;
    private boolean[][] flagged;
    private int remainingTilesToClear;
    private int mines;

    private int gameState;

    public GameSession() {
        map = new int[HEIGHT][];
        covered = new boolean[HEIGHT][];
        flagged = new boolean[HEIGHT][];

        for (int c = 0; c < HEIGHT; ++c) {

            map[c] = new int[WIDTH];
            covered[c] = new boolean[WIDTH];
            flagged[c] = new boolean[WIDTH];

            for (int d = 0; d < WIDTH; ++d) {
                covered[c][d] = true;
                flagged[c][d] = false;
                map[c][d] = POSITION_BLANK;
            }
        }
        remainingTilesToClear = WIDTH * HEIGHT;
    }

    public void placeRandomMines(int n) {

        int x;
        int y;
        Random random = new Random();
        mines = n;
        while (n > 0) {

            x = random.nextInt(WIDTH - 2) + 1;
            y = random.nextInt(HEIGHT - 2) + 1;

            if (map[y][x] != POSITION_MINE) {

                --n;
                map[y][x] = POSITION_MINE;
            }
        }


        placeNumbersOnBoard();
    }

    private void placeNumbersOnBoard() {
        for (int y = 1; y < getHeight() - 1; ++y) {
            for (int x = 1; x < getWidth() - 1; ++x) {
                if (map[y][x] == POSITION_MINE) {

                    if (map[y - 1][x - 1] != POSITION_MINE)
                        ++map[y - 1][x - 1];

                    if (map[y][x - 1] != POSITION_MINE)
                        ++map[y][x - 1];

                    if (map[y + 1][x - 1] != POSITION_MINE)
                        ++map[y + 1][x - 1];

                    if (map[y - 1][x] != POSITION_MINE)
                        ++map[y - 1][x];

                    if (map[y + 1][x] != POSITION_MINE)
                        ++map[y + 1][x];

                    if (map[y - 1][x + 1] != POSITION_MINE)
                        ++map[y - 1][x + 1];

                    if (map[y][x + 1] != POSITION_MINE)
                        ++map[y][x + 1];

                    if (map[y + 1][x + 1] != POSITION_MINE)
                        ++map[y + 1][x + 1];
                }
            }
        }
    }

    public boolean isCoveredAt(int x, int y) {

        if (x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT)
            return true;

        return covered[y][x];
    }

    public int getWidth() {

        return WIDTH;
    }

    public int getHeight() {

        return HEIGHT;
    }

    public int getPos(int x, int y) {

        if (x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT)
            return POSITION_BLANK;

        return map[y][x];
    }

    public void poke(int x, int y) {

        if (x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT)
            return;

        if (flagged[x][y]) {

            flagged[y][x] = false;
            return;
        }


        switch (map[y][x]) {
            case POSITION_MINE:

                map[y][x] = GameSession.POSITION_MINE_POKED;
                gameState = GameSession.GAMESTATE_GAMEOVER;
                return;

            case POSITION_BLANK:
                floodUncover(x, y);

                if (covered[y][x] && !flagged[y][x])
                    this.remainingTilesToClear--;

                covered[y][x] = false;
            default:

                if (covered[y][x])
                    this.remainingTilesToClear--;

                covered[y][x] = false;
        }

        if (this.remainingTilesToClear == mines)
            gameState = GameSession.GAMESTATE_FININSHED;


    }


    private void floodUncover(int x, int y) {
        if (covered[y][x] && !flagged[y][x]) {

            if (map[y][x] == POSITION_BLANK) {

                if (covered[y][x])
                    this.remainingTilesToClear--;

                covered[y][x] = false;

                if (x > 0)
                    floodUncover(x - 1, y);

                if ((x + 1) < WIDTH)
                    floodUncover(x + 1, y);

                if (y > 0)
                    floodUncover(x, y - 1);

                if ((y + 1) < HEIGHT)
                    floodUncover(x, y + 1);

                if (x > 0 && y > 0)
                    floodUncover(x - 1, y - 1);

                if ((x + 1) < WIDTH && (y + 1) < HEIGHT)
                    floodUncover(x + 1, y + 1);

                if (y > 0 && (x + 1) < WIDTH)
                    floodUncover(x + 1, y - 1);

                if ((y + 1) < HEIGHT && x > 0)
                    floodUncover(x - 1, y + 1);


            }
            if (map[y][x] != POSITION_MINE) {

                if (covered[y][x])
                    this.remainingTilesToClear--;

                covered[y][x] = false;
            }
        }
    }

    public boolean isFinished() {

        return gameState == GameSession.GAMESTATE_GAMEOVER || isVictory();
    }

    public boolean isVictory() {

        return gameState == GameSession.GAMESTATE_FININSHED;
    }

    public void uncoverAt(int x, int y) {

        if (isValid(x, y)) {
            covered[y][x] = false;
            flagged[y][x] = false;
        }
    }

    public void flag(int x, int y) {

        if (isValid(x, y))
            flagged[y][x] = !flagged[y][x];
    }

    public boolean isFlaggedAt(int x, int y) {

        if (isValid(x, y))
            return flagged[y][x];
        else
            return false;
    }

    private boolean isValid(int x, int y) {

        if (x < 0)
            return false;

        if (y < 0)
            return false;

        if (x >= WIDTH)
            return false;

        return y < HEIGHT;
    }

    public void saveState(OutputStream os) throws IOException {

        DataOutputStream dos = new DataOutputStream(os);

        for (int c = 0; c < HEIGHT; ++c) {

            for (int d = 0; d < WIDTH; ++d) {
                dos.writeBoolean(covered[c][d]);
                dos.writeBoolean(flagged[c][d]);
                dos.writeInt(map[c][d]);
            }
        }
        dos.writeInt(remainingTilesToClear);
    }

    public void loadState(InputStream is) throws IOException {

        DataInputStream dis = new DataInputStream(is);

        for (int c = 0; c < HEIGHT; ++c) {

            for (int d = 0; d < WIDTH; ++d) {

                covered[c][d] = dis.readBoolean();
                flagged[c][d] = dis.readBoolean();
                map[c][d] = dis.readInt();
            }
        }
        remainingTilesToClear = dis.readInt();
    }
}

File: SonarMinefield/app/src/main/java/br/odb/sonarminefield/GameSession.kt
package br.odb.sonarminefield

import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.InputStream
import java.io.OutputStream
import java.util.*

/**
 * @author monty
 */
class GameSession {
    private val map: Array<IntArray?>
    private val covered: Array<BooleanArray?>
    private val flagged: Array<BooleanArray?>
    private var remainingTilesToClear: Int
    private var mines = 0

    val width = 10
    val height = 15

    companion object {
        const val POSITION_MINE = 9
        const val POSITION_COVERED = 10
        const val POSITION_MINE_POKED = 11
        const val POSITION_FLAGGED = 12
        const val POSITION_BLANK = 0
    }

    fun placeRandomMines(n: Int) {
        var n = n
        var x: Int
        var y: Int
        val random = Random()
        mines = n
        while (n > 0) {
            x = random.nextInt(width - 2) + 1
            y = random.nextInt(height - 2) + 1
            if (map[y]!![x] != POSITION_MINE) {
                --n
                map[y]!![x] = POSITION_MINE
            }
        }
        placeNumbersOnBoard()
    }

    private fun placeNumbersOnBoard() {
        for (y in 1 until height - 1) {
            for (x in 1 until width - 1) {
                if (map[y]!![x] == POSITION_MINE) {
                    if (map[y - 1]!![x - 1] != POSITION_MINE) ++map[y - 1]!![x - 1]
                    if (map[y]!![x - 1] != POSITION_MINE) ++map[y]!![x - 1]
                    if (map[y + 1]!![x - 1] != POSITION_MINE) ++map[y + 1]!![x - 1]
                    if (map[y - 1]!![x] != POSITION_MINE) ++map[y - 1]!![x]
                    if (map[y + 1]!![x] != POSITION_MINE) ++map[y + 1]!![x]
                    if (map[y - 1]!![x + 1] != POSITION_MINE) ++map[y - 1]!![x + 1]
                    if (map[y]!![x + 1] != POSITION_MINE) ++map[y]!![x + 1]
                    if (map[y + 1]!![x + 1] != POSITION_MINE) ++map[y + 1]!![x + 1]
                }
            }
        }
    }

    fun isCoveredAt(x: Int, y: Int): Boolean {
        return if (x < 0 || x >= width || y < 0 || y >= height) true else covered[y]!![x]
    }

    fun getPos(x: Int, y: Int): Int {
        return if (x < 0 || x >= width || y < 0 || y >= height) POSITION_BLANK else map[y]!![x]
    }

    fun poke(x: Int, y: Int) : GameBoard.GameOutcome {

        if (x < 0 || x >= width || y < 0 || y >= height) return GameBoard.GameOutcome.kPlaying

        if (flagged[y]!![x]) {
            flagged[y]!![x] = false
            return GameBoard.GameOutcome.kPlaying
        }
        when (map[y]!![x]) {
            POSITION_MINE -> {
                map[y]!![x] = POSITION_MINE_POKED
                return GameBoard.GameOutcome.kLost
            }
            POSITION_BLANK -> {
                floodUncover(x, y)
                if (covered[y]!![x] && !flagged[y]!![x]) remainingTilesToClear--
                covered[y]!![x] = false
                if (covered[y]!![x]) remainingTilesToClear--
                covered[y]!![x] = false
            }
            else -> {
                if (covered[y]!![x]) remainingTilesToClear--
                covered[y]!![x] = false
            }
        }
        if (remainingTilesToClear == mines) return GameBoard.GameOutcome.kWon


        return GameBoard.GameOutcome.kPlaying
    }

    private fun floodUncover(x: Int, y: Int) {
        if (covered[y]!![x] && !flagged[y]!![x]) {
            if (map[y]!![x] == POSITION_BLANK) {
                if (covered[y]!![x]) remainingTilesToClear--
                covered[y]!![x] = false
                if (x > 0) floodUncover(x - 1, y)
                if (x + 1 < width) floodUncover(x + 1, y)
                if (y > 0) floodUncover(x, y - 1)
                if (y + 1 < height) floodUncover(x, y + 1)
                if (x > 0 && y > 0) floodUncover(x - 1, y - 1)
                if (x + 1 < width && y + 1 < height) floodUncover(x + 1, y + 1)
                if (y > 0 && x + 1 < width) floodUncover(x + 1, y - 1)
                if (y + 1 < height && x > 0) floodUncover(x - 1, y + 1)
            }
            if (map[y]!![x] != POSITION_MINE) {
                if (covered[y]!![x]) remainingTilesToClear--
                covered[y]!![x] = false
            }
        }
    }

    fun uncoverAt(x: Int, y: Int) {
        if (isValid(x, y)) {
            covered[y]!![x] = false
            flagged[y]!![x] = false
        }
    }

    fun flag(x: Int, y: Int) {
        if (isValid(x, y)) flagged[y]!![x] = !flagged[y]!![x]
    }

    fun isFlaggedAt(x: Int, y: Int): Boolean {
        return if (isValid(x, y)) flagged[y]!![x] else false
    }

    private fun isValid(x: Int, y: Int): Boolean {
        if (x < 0) return false
        if (y < 0) return false
        return if (x >= width) false else y < height
    }

    fun saveState(outputStream: OutputStream?) {
        val dos = DataOutputStream(outputStream)
        for (c in 0 until height) {
            for (d in 0 until width) {
                dos.writeBoolean(covered[c]!![d])
                dos.writeBoolean(flagged[c]!![d])
                dos.writeInt(map[c]!![d])
            }
        }
        dos.writeInt(remainingTilesToClear)
    }

    fun loadState(inputStream: InputStream?) {
        val dis = DataInputStream(inputStream)
        for (c in 0 until height) {
            for (d in 0 until width) {
                covered[c]!![d] = dis.readBoolean()
                flagged[c]!![d] = dis.readBoolean()
                map[c]!![d] = dis.readInt()
            }
        }
        remainingTilesToClear = dis.readInt()
    }

    fun clearBorders() {
        for (c in 0 until height) {
            poke(0, c)
            poke(width - 1, c)
        }
        for (d in 0 until width) {
            poke(d, 0)
            poke(d, height - 1)
        }
    }


    init {
        map = arrayOfNulls(height)
        covered = arrayOfNulls(height)
        flagged = arrayOfNulls(height)
        for (c in 0 until height) {
            map[c] = IntArray(width)
            covered[c] = BooleanArray(width)
            flagged[c] = BooleanArray(width)
            for (d in 0 until width) {
                covered[c]!![d] = true
                flagged[c]!![d] = false
                map[c]!![d] = POSITION_BLANK
            }
        }
        remainingTilesToClear = width * height
    }
}